kube-derive/src/cel_schema.rs

```diff
@@ -0,0 +1,238 @@
+    }
+
+    fn default_schemars() -> Path {
+        parse_quote! { ::schemars }
+    }
+
+    fn default_serde() -> Path {
+        parse_quote! { ::serde }
+    }
+}
+
+pub(crate) fn derive_validated_schema(input: TokenStream) -> TokenStream {
+    let mut ast: DeriveInput = match syn::parse2(input) {
+        Err(err) => return err.to_compile_error(),
+        Ok(di) => di,
+    };
+
+    let CELSchema {
+        crates: Crates {
+            kube_core,
+            schemars,
+            serde,
+        },
+        ident,
+        rules,
+    } = match CELSchema::from_derive_input(&ast) {
+        Err(err) => return err.write_errors(),
+        Ok(attrs) => attrs,
+    };
+
+    // Collect global structure validation rules
+    let struct_name = ident.to_string();
+    let struct_rules: Vec<TokenStream> = rules.iter().map(|r| quote! {#r,}).collect();
+
+    // Remove all unknown attributes
+    // Has to happen on the original definition at all times, as we don't have #[derive] stanzes.
+    let attribute_whitelist = ["serde", "schemars", "doc"];
+    ast.attrs = ast
+        .attrs
+        .iter()
+        .filter(|attr| attribute_whitelist.iter().any(|i| attr.path().is_ident(i)))
+        .cloned()
+        .collect();
+
+    let struct_data = match ast.data {
+        syn::Data::Struct(ref mut struct_data) => struct_data,
+        _ => return quote! {},
+    };
+
+    // Preserve all serde attributes, to allow #[serde(rename_all = "camelCase")] or similar
+    let struct_attrs: Vec<TokenStream> = ast.attrs.iter().map(|attr| quote! {#attr}).collect();
+    let mut property_modifications = vec![];
+    if let syn::Fields::Named(fields) = &mut struct_data.fields {
+        for field in &mut fields.named {
+            let Rule { rules, .. } = match Rule::from_field(field) {
+                Ok(rule) => rule,
+                Err(err) => return err.write_errors(),
+            };
+
+            // Remove all unknown attributes
+            // Has to happen on the original definition at all times, as we don't have #[derive] stanzes.
+            field.attrs = field
+                .attrs
+                .iter()
+                .filter(|attr| attribute_whitelist.iter().any(|i| attr.path().is_ident(i)))
+                .cloned()
+                .collect();
+
+            if rules.is_empty() {
+                continue;
+            }
+
+            let rules: Vec<TokenStream> = rules.iter().map(|r| quote! {#r,}).collect();
+
+            // We need to prepend derive macros, as they were consumed by this macro processing, being a derive by itself.
+            property_modifications.push(quote! {
+                {
+                    #[derive(#serde::Serialize, #schemars::JsonSchema)]
+                    #(#struct_attrs)*
+                    #[automatically_derived]
+                    #[allow(missing_docs)]
+                    struct Validated {
+                        #field
+                    }
```

Comment:
So basically, this creates a validated wrapper struct for EVERY field that uses `#[cel_validate]`, and these structs do not clash because they are inlined in a private scope for the purposes of using `merge_properties` only?

Reply:
Yeah, they are in their own code block, so they don't match by schema id, even though have the same name. We just take the property from the struct before it goes out of scope. And since there is only one field, there is no need in ordering or knowing the renamed field name with "serde" before performing replacement.